test/Runtime/type_discovery.swift

```diff
@@ -0,0 +1,22 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+protocol P { }
+struct A: P { }
+class B: P { }
+enum C: P { }
+actor D: P { }
+struct E /* does not conform */ { }
+
+var typeList = [P.Type]()
+
+_enumerateTypes(fromImageAt: #dsohandle, conformingTo: P.self) { type, _ in
+  typeList.append(type as! P.Type)
```

Comment:
If this force casts fails then the test process will halt. I'd suggest an optional cast and adding it only if it's non-nil instead:

```swift 
if let casted = type as? P.Type { 
    typeList.append(casted)
}
```

This will still give us a clear error message as well, since any unexpected type issues will cause the array to be empty.

Reply:
It's impossible for it to fail—it's only a constraint of the language that the value isn't already a `P.Type`. So if this cast fails, it _should_ crash.